src/litdata/streaming/item_loader.py

```diff
@@ -85,14 +97,19 @@ class PyTreeLoader(BaseItemLoader):
     def __init__(self) -> None:
         self._chunk_filepaths: Dict[str, bool] = {}
 
-    def generate_intervals(self) -> List[Tuple[int, int]]:
+    def generate_intervals(self) -> List[Tuple[int, int, int, int]]:
```

Comment:
I think we could have avoided to leak sub-sampling entirely into the item loader by getting the intervals and then applying sub-sampling to them.

Reply:
While saving the dataset's state dict, we might lose information on which are the relevant intervals. 
```python
"item_loader": self.item_loader.state_dict() if self.item_loader else None,
```
And only TokenLoader saves `block_size` info. But, we can modify this.

---

Also, let's say a user subsampled a dataset (let's say 0.3), and after some EDA, she wants to train a dummy model. So, she might want to split the subsampled dataset into (0.9, 0.1) and do her training and then validate.

To support this (`split a subsampled dataset`), we need to tweak the `roi` and `subsampled_files` only. Rest all can remain as it is, and we are done with deepcopy.

If these properties are part of StreamingDataset class, we can directly access, else, we will be required to go deeper and then get the item_loader and then it's roi and relevant filenames.

I feel like, I'm trying hard not to change my code anymore, lol. But, that's not the case. This is what I genuinely feel.🫠

Reply:
Here's a new testcase added for the same:
```python
def test_split_a_subsampled_dataset(tmpdir, compression):
    cache = Cache(str(tmpdir), chunk_size=15, compression=compression)
    for i in range(1000):
        cache[i] = i
    cache.done()
    cache.merge()

    _sub_sampled_streaming_dataset = StreamingDataset(input_dir=str(tmpdir), subsample=0.3)

    assert len(_sub_sampled_streaming_dataset) == 300 # 1000 * 0.3

    _split_fraction = [0.2, 0.3, 0.4, 0.1]

    split_datasets = train_test_split(_sub_sampled_streaming_dataset, _split_fraction)

    assert all(len(split_datasets[i])==int(300*split) for i, split in enumerate(_split_fraction))
```

Reply:
Ok, sounds good with me.